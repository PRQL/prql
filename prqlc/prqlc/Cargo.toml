[package]
description = "PRQL is a modern language for transforming data â€” a simple, powerful, pipelined SQL replacement."
name = "prqlc"

edition.workspace = true
license.workspace = true
repository.workspace = true
rust-version.workspace = true
version.workspace = true

# Required for `cargo-msrv`, which doesn't yet support workspaces
metadata.msrv = "1.70.0"

[features]
cli = [
  "anyhow",
  "clap_complete_command",
  "clap_complete",
  "clap",
  "clio",
  "color-eyre",
  "colorchoice-clap",
  "env_logger",
  "is-terminal",
  "minijinja",
  "notify",
  "serde_yaml",
  "walkdir",
]
default = ["cli"]
serde_yaml = [
  "prqlc-ast/serde_yaml",
  "prqlc-parser/serde_yaml",
  "dep:serde_yaml",
]
# Technically tokio could be limited to external tests, but its types are in
# signatures which would require lots of conditional compilation.
test-dbs = ["anyhow", "duckdb", "glob", "rusqlite", "tokio"]
test-dbs-external = ["anyhow", "duckdb", "glob", "mysql", "pg_bigdecimal", "postgres", "rusqlite", "tiberius", "tokio", "tokio-util"]

[dependencies]
prqlc-ast = {path = "../prqlc-ast", version = "0.11.5"}
prqlc-parser = {path = "../prqlc-parser", version = "0.11.5"}

anstream = {version = "0.6.14", features = ["auto"]}
ariadne = "0.4.1"
chrono = "0.4.38"
csv = "1.3.0"
enum-as-inner = "0.6.0"
itertools = {workspace = true}
log = {workspace = true}
once_cell = "1.19.0"
regex = "1.10.4"
semver = {version = "1.0.23", features = ["serde"]}
# serde is required for the `from_text` feature of PRQL, so we can't really put
# it behind a feature.
serde = {workspace = true}
serde_json = {workspace = true}
serde_yaml = {workspace = true, optional = true}
sqlformat = "0.2.3"
sqlparser = {version = "0.46.0", features = ["serde"]}
strum = {version = "0.26.2", features = ["std", "derive"]}
strum_macros = "0.26.2"

[target.'cfg(not(target_family="wasm"))'.dependencies]

# unique dependencies from the CLI, marked as optional and included in the 'cli'
# feature
anyhow = {version = "1.0.86", features = ["backtrace"], optional = true}
clap = {version = "4.4.18", features = ["derive", "env", "wrap_help"], optional = true}
clap_complete_command = {version = "0.5.1", optional = true}
clio = {version = "0.3.3", features = ['clap-parse'], optional = true}
color-eyre = {version = "0.6.3", optional = true}
colorchoice-clap = {version = "1.0.0", optional = true}
env_logger = {version = "0.10.2", features = ["color"], optional = true}
is-terminal = {version = "0.4.12", optional = true}
notify = {version = "6.1.1", optional = true}
walkdir = {version = "2.5.0", optional = true}

# Not direct dependencies, but pinning because of bugs in previous versions. Can
# remove when dependencies no longer use it. (If CI passes, it can be removed.)
clap_complete = {version = "4.4.9", optional = true}

# We use minijinja just for the Jinja lexer, which is not part of the
# public interface which is covered by semver guarantees.
minijinja = {version = "2.0.1", features = ["unstable_machinery"], optional = true}

# For integration tests. These are gated by the `test-dbs` and `test-dbs-external` features,
# rather than dev-dependencies, because dev-dependencies can't be optional.

duckdb = {version = "0.10.2", optional = true, features = ["bundled", "chrono"]}
glob = {version = "0.3.1", optional = true}
mysql = {version = "25", optional = true}
pg_bigdecimal = {version = "0.1.0", optional = true}
postgres = {version = "0.19.7", optional = true}
rusqlite = {version = "0.31.0", optional = true, features = ["bundled", "csvtab"]}
tiberius = {version = "0.12.2", optional = true, default-features = false, features = ["sql-browser-tokio", "bigdecimal", "time", "rustls", "tds73"]}
tokio = {version = "1.37.0", optional = true, features = ["full"]}
tokio-util = {version = "0.7", optional = true, features = ["compat"]}

[dev-dependencies]
cfg-if = "1.0"
insta = {workspace = true}
insta-cmd = {workspace = true}
rstest = "0.19.0"
similar = {version = "2.5.0"}
similar-asserts = "1.5.0"
test_each_file = "0.3.2"

[target.'cfg(not(target_family="wasm"))'.dev-dependencies]
criterion = {version = "0.5.1"}

[[bench]]
harness = false
name = "bench"

# Putting these pre-release replacements in the workspace root causes it to
# refer to the path relative to `prqlc`, so we choose this crate out of
# default.

[[package.metadata.release.pre-release-replacements]]
exactly = 1
file = "../../web/book/src/project/target.md"
replace = 'prql version:"{{version}}"'
search = 'prql version:"[\d.]+"'

[[package.metadata.release.pre-release-replacements]]
exactly = 1
file = "../../CHANGELOG.md"
replace = "# PRQL Changelog\n\n## [unreleased]\n\n**Language**:\n\n**Features**:\n\n**Fixes**:\n\n**Documentation**:\n\n**Web**:\n\n**Integrations**:\n\n**Internal changes**:\n\n**New Contributors**:\n\n"
search = "# PRQL Changelog\n\n"
