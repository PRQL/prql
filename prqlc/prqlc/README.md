# PRQL compiler

`prqlc` is the reference implementation of a compiler from PRQL to SQL, written
in Rust. It also serves as the CLI.

For more on PRQL, check out the [PRQL website](https://prql-lang.org) or the
[PRQL repo](https://github.com/PRQL/prql).

## CLI

`prqlc` serves as a CLI for the PRQL compiler. It is a single, dependency-free
binary that compiles PRQL into SQL.

## Usage

### `prqlc compile`

This command works as a filter that compiles a PRQL string into an SQL string.

```sh
$ echo 'from employees | filter has_dog | select salary' | prqlc compile

SELECT
  salary
FROM
  employees
WHERE
  has_dog
```

A PRQL query can be executed with CLI tools compatible with SQL,, such as
[DuckDB CLI](https://duckdb.org/docs/api/cli.html).

```sh
$ curl -fsL https://raw.githubusercontent.com/PRQL/prql/0.8.1/prql-compiler/tests/integration/data/chinook/albums.csv -o albums.csv
$ echo 'from `albums.csv` | take 3' | prqlc compile | duckdb
┌──────────┬───────────────────────────────────────┬───────────┐
│ album_id │                 title                 │ artist_id │
│  int64   │                varchar                │   int64   │
├──────────┼───────────────────────────────────────┼───────────┤
│        1 │ For Those About To Rock We Salute You │         1 │
│        2 │ Balls to the Wall                     │         2 │
│        3 │ Restless and Wild                     │         2 │
└──────────┴───────────────────────────────────────┴───────────┘
```

Executing this command without any argument will start interactive mode,
allowing a PRQL query to be written interactively. In this mode, after writing
PRQL and press `Ctrl-d` (Linux, macOS) or `Ctrl-z` (Windows) to display the
compiled SQL.

```sh
prqlc compile
```

Just like when using it as a filter, SQL string output can be passed to the
DuckDB CLI and similar tools.

```sh
$ prqlc compile | duckdb
Enter PRQL, then press ctrl-d to compile:

from `albums.csv`
take 3
┌──────────┬───────────────────────────────────────┬───────────┐
│ album_id │                 title                 │ artist_id │
│  int64   │                varchar                │   int64   │
├──────────┼───────────────────────────────────────┼───────────┤
│        1 │ For Those About To Rock We Salute You │         1 │
│        2 │ Balls to the Wall                     │         2 │
│        3 │ Restless and Wild                     │         2 │
└──────────┴───────────────────────────────────────┴───────────┘
```

## Installation

[![Packaging status](https://repology.org/badge/vertical-allrepos/prqlc.svg)](https://repology.org/project/prqlc/versions)

### via Homebrew (macOS, Linux)

```sh
brew install prqlc
```

### via winget (Windows)

```sh
winget install prqlc
```

### From GitHub release page

Precompiled binaries are available for Linux, macOS, and Windows on the
[PRQL release page](https://github.com/PRQL/prql/releases).

### From source

```sh
# From crates.io
cargo install prqlc
```

```sh
# From a local PRQL repository
cargo install --path prqlc/prqlc
```

### Shell completions

The `prqlc shell-completion` command prints a shell completion script for
supported shells, and saving the printed scripts to files makes for shells to
load completions for each session.

#### Bash

For Linux:

```sh
prqlc shell-completion bash >/etc/bash_completion.d/prqlc
```

For macOS:

```sh
prqlc shell-completion bash >/usr/local/etc/bash_completion.d/prqlc
```

#### fish

```sh
prqlc shell-completion fish >~/.config/fish/completions/prqlc.fish
```

#### PowerShell

```powershell
mkdir -Path (Split-Path -Parent $profile) -ErrorAction SilentlyContinue
prqlc shell-completion powershell >path/to/prqlc.ps1
echo 'Invoke-Expression -Command path/to/prqlc.ps1' >>$profile
```

#### zsh

```sh
prqlc shell-completion zsh >"${fpath[1]}/_prqlc"
```

Ensure that the following lines are present in `~/.zshrc`:

```sh
autoload -U compinit
compinit -i
```

## Helpers

Cheat sheets for `prqlc` are available on various websites and with various
tools.

- [`tldr`](https://tldr.sh/)
  ([on the web](https://tldr.inbrowser.app/pages/common/prqlc))
- [`eg`](https://github.com/srsudar/eg)

<!-- Issues: #2034 cheat/cheatsheets, #2041 devhints.io -->

## Library

For more usage examples and the library documentation, check out the
[`prqlc` documentation](https://docs.rs/prqlc/).

### Library installation

```shell
cargo add prqlc
```

### Examples

Compile a PRQL string to a SQLite dialect string:

```rust
// In a file src/main.rs

use prqlc::{compile, Options, DisplayOptions, Target, sql::Dialect};

let prql = "from employees | select {name, age}";
let opts = &Options {
    format: false,
    target: Target::Sql(Some(Dialect::SQLite)),
    signature_comment: false,
    display: DisplayOptions::Plain,
    ..Default::default()
};
let sql = compile(&prql, opts).unwrap();
assert_eq!("SELECT name, age FROM employees", sql);
```
