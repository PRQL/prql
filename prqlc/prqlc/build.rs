use std::error::Error;

pub fn main() -> Result<(), Box<dyn Error>> {
    Ok(())
}
