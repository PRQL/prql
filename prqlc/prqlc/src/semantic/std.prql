# The PRQL standard library defines the following functions and transforms.
# The definitions are whitespace insensitive, and have this form:
#
# ```
# let my_func = param1 param2 ...  -> <return_type> body_expr
# ```
#
# Where:
#   * `my_func` is the name of the function
#   * `param1` is the first parameter optionally followed by a type in "< ... >"
#   * `param2` etc. follow the same pattern as param1
#   * `<return_type>` is the type of result wrapped in "< ... >"
#   * `body_expr` defines the function body that creates the result.
#     It can be PRQL code or `internal ...` to indicate internal compiler code.

# Operators

let mul = left right -> <int> internal std.mul
let div_i = left right -> <int> internal std.div_i
let div_f = left right -> <int> internal std.div_f
let mod = left right -> <int> internal std.mod
let add = left right -> <int> internal std.add
let sub = left right -> <int> internal std.sub
let eq = left right -> <bool> internal std.eq
let ne = left right -> <bool> internal std.ne
let gt = left right -> <bool> internal std.gt
let lt = left right -> <bool> internal std.lt
let gte = left right -> <bool> internal std.gte
let lte = left right -> <bool> internal std.lte
let and = left<bool> right<bool> -> <bool> internal std.and
let or = left<bool> right<bool> -> <bool> internal std.or
let coalesce = left right -> internal std.coalesce
let regex_search = text pattern -> <bool> internal std.regex_search

let neg = expr<int> -> <int> internal std.neg
let not = expr<bool> -> <bool> internal std.not

# Types

## Type primitives
type int = int
type float = float
type bool = bool
type text = text
type date = date
type time = time
type timestamp = timestamp
type `func` = func

## Generic array

## Range
# type range <T> = {start = T, end = T}

## Transform
# type transform <I, O> = func [I] -> [O]

# Functions

## Relational transforms
let from = func <R>
  tbl <R>
  -> <R> tbl

@(coerce_tuple 0)
@(implicit_closure 0 this:1)
let select = func <I, A>
  assigns <A>
  tbl <[I]>
  -> <[A]> internal select

@(implicit_closure 0 this:1)
let filter = func <T>
  condition <bool>
  tbl <[T]>
  -> <[T]> internal filter

@(coerce_tuple 0)
@(implicit_closure 0 this:1)
let derive = func <I, A>
  assigns <A>
  tbl <[I]>
  -> <[{I, ..A}]> internal derive

@(coerce_tuple 0)
@(implicit_closure 0 this:1)
let aggregate = func <I, A>
  assigns <A>
  tbl <[I]>
  -> <[A]> internal aggregate

@(coerce_tuple 0)
@(implicit_closure 0 this:1)
let sort = func <I>
  by
  tbl <[I]>
  -> <[I]> internal sort

let take = func <I>
  expr
  tbl <[I]>
  -> <[I]> internal take

@(implicit_closure 1 this:2 that:0)
let join = func <A, B>
  with <[B]>
  condition <bool>
  `noresolve.side`:inner
  tbl <[A]>
  -> <[{A, B}]> internal join

@(coerce_tuple 0)
@(implicit_closure 0 this:2)
let group = func <I, B, PO>
  by<B>
  pipeline <func [I - B] -> [PO]>
  tbl <[I]>
  -> <[{B, ..PO}]> internal group

let window = func
  rows:0..-1
  range:0..-1
  expanding <bool>:false
  rolling <int>:0
  pipeline
  tbl
  -> internal window

let append = bottom top -> internal append
# let intersect = bottom top -> (
#   t = top
#   join (b = bottom) (std.tuple_every (std.tuple_map std._eq (std.tuple_zip t.* b.*)))
#   select t.*
# )
# let remove = bottom top -> (
#   t = top
#   join side:left (b = bottom) (std.tuple_every (std.tuple_map std._eq (std.tuple_zip t.* b.*)))
#   filter (std.tuple_every (std.tuple_map std._is_null b.*))
#   select t.*
# )
let loop = func
  pipeline
  top
  -> internal loop

## Aggregate functions
# These return either a scalar when used within `aggregate`, or a column when used anywhere else.

let min = column -> <int> internal std.min

let max = column -> <int> internal std.max

let sum = column -> <int> internal std.sum

let average = column -> <float> internal std.average

let stddev = column -> <float> internal std.stddev

let all = column -> <bool> internal std.all

let any = column -> <bool> internal std.any

let concat_array = column -> <text> internal std.concat_array

# Counts number of items in the column.
# Note that the count will include null values.
let count = colum -> <int> internal count

# Deprecated in favour of filterning input to the [std.count] function (not yet implemented).
@{deprecated}
let count_distinct = column -> internal std.count_distinct

## Window functions
let lag =   offset <int>    column -> internal std.lag
let lead =  offset <int>    column -> internal std.lead
let first      = column -> internal std.first
let last       = column -> internal std.last
let rank       = column -> internal std.rank
let rank_dense = column -> internal std.rank_dense
let row_number = column -> internal row_number

# Mathematical functions
module math {
  let abs = column -> <int> internal std.math.abs
  let floor = column -> <int> internal std.math.floor
  let ceil = column -> <int> internal std.math.ceil
  let pi <float> = internal std.math.pi
  let exp = column -> <int> internal std.math.exp
  let ln = column -> <int> internal std.math.ln
  let log10 = column -> <int> internal std.math.log10
  let log = base<int> column -> <int> internal std.math.log
  let sqrt = column -> <int> internal std.math.sqrt
  let degrees = column -> <int> internal std.math.degrees
  let radians = column -> <int> internal std.math.radians
  let cos = column -> <int> internal std.math.cos
  let acos = column -> <int> internal std.math.acos
  let sin = column -> <int> internal std.math.sin
  let asin = column -> <int> internal std.math.asin
  let tan = column -> <int> internal std.math.tan
  let atan = column -> <int> internal std.math.atan
  let pow = exponent<int> column -> <int> internal std.math.pow
  let round = n_digits column -> internal std.math.round
}

## Misc functions
let as = `noresolve.type` column -> internal std.as
let in = pattern value -> <bool> internal in

## Tuple functions
let tuple_every = func list -> <bool> internal tuple_every
let tuple_map = func fn <func> list -> internal tuple_map
let tuple_zip = func a b -> internal tuple_zip
let _eq = func a -> internal _eq
let _is_null = func a -> a == null

## Misc
let from_text = func <R: {}> input<text> `noresolve.format`:csv -> <[R]> internal from_text

## Text functions
module text {
  let lower = column -> <text> internal std.text.lower
  let upper = column -> <text> internal std.text.upper
  let ltrim = column -> <text> internal std.text.ltrim
  let rtrim = column -> <text> internal std.text.rtrim
  let trim = column -> <text> internal std.text.trim
  let length = column -> <int> internal std.text.length
  let extract = offset<int> length<int> column -> <text> internal std.text.extract
  let replace = pattern<text> replacement<text> column -> <text> internal std.text.replace
  let starts_with = prefix<text> column -> <bool> internal std.text.starts_with
  let contains = substr<text> column -> <bool> internal std.text.contains
  let ends_with = suffix<text> column -> <bool> internal std.text.ends_with
}

## Date functions
module date {
  let to_text = format<text> column -> <text> internal std.date.to_text
}

## File-reading functions, primarily for DuckDB
let read_parquet = func <R: {}> source<text> -> <[R]> internal std.read_parquet
# let read_csv = func <R: {}> source<text> -> <[R]> internal std.read_csv

## PRQL compiler functions
module `prql` {
  let version <text> = internal prql_version
}

# Deprecated, will be removed in 0.12.0
let prql_version <text> = internal prql_version
