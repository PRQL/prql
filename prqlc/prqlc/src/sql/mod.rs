//! Backend for translating RQ into SQL

mod dialect;
mod gen_expr;
mod gen_projection;
mod gen_query;
mod keywords;
mod operators;
mod pq;

pub use dialect::{Dialect, SupportLevel};
pub use pq::ast as pq_ast;

use self::dialect::DialectHandler;
use self::pq::ast::Cte;
use self::pq::context::AnchorContext;
use crate::debug;
use crate::ir::rq;
use crate::Result;
use crate::{compiler_version, Options};

/// Translate a PRQL AST into a SQL string.
pub fn compile(query: rq::RelationalQuery, options: &Options) -> Result<String> {
    let crate::Target::Sql(dialect) = options.target;
    let sql_ast = gen_query::translate_query(query, dialect)?;

    let sql = sql_ast.to_string();

    // formatting
    let sql = if options.format {
        let formatted = sqlformat::format(
            &sql,
            &sqlformat::QueryParams::default(),
            &sqlformat::FormatOptions::default(),
        );

        formatted + "\n"
    } else {
        sql
    };

    debug::log_entry(|| debug::DebugEntryKind::ReprSql(sql.clone()));

    // signature
    let sql = if options.signature_comment {
        let pre = if options.format { "\n" } else { " " };
        let post = if options.format { "\n" } else { "" };
        let target = dialect
            .map(|d| format!("target:sql.{d} "))
            .unwrap_or_default();
        let signature = format!(
            "{pre}-- Generated by PRQL compiler version:{} {}(https://prql-lang.org){post}",
            compiler_version(),
            target,
        );
        sql + &signature
    } else {
        sql
    };

    Ok(sql)
}

#[derive(Debug)]
struct Context {
    pub dialect: Box<dyn DialectHandler>,
    pub dialect_enum: Dialect,

    pub anchor: AnchorContext,

    // stuff regarding current query
    query: QueryOpts,

    // stuff regarding parent queries
    query_stack: Vec<QueryOpts>,

    pub ctes: Vec<Cte>,
}

#[derive(Clone, Debug)]
struct QueryOpts {
    /// When true, column references will not include table names prefixes.
    pub omit_ident_prefix: bool,

    /// True iff codegen should generate expressions before SELECT's projection is applied.
    /// For example:
    /// - WHERE needs `pre_projection=true`, but
    /// - ORDER BY needs `pre_projection=false`.
    pub pre_projection: bool,

    /// When false, queries will contain nested sub-queries instead of WITH CTEs.
    pub allow_ctes: bool,

    /// When false, * are not allowed.
    pub allow_stars: bool,

    /// True when translating function that will have an OVER clause.
    pub window_function: bool,
}

impl Default for QueryOpts {
    fn default() -> Self {
        QueryOpts {
            omit_ident_prefix: false,
            pre_projection: false,
            allow_ctes: true,
            allow_stars: true,
            window_function: false,
        }
    }
}

impl Context {
    fn new(dialect: Dialect, anchor: AnchorContext) -> Self {
        Context {
            dialect: dialect.handler(),
            dialect_enum: dialect,
            anchor,
            query: QueryOpts::default(),
            query_stack: Vec::new(),
            ctes: Vec::new(),
        }
    }

    fn push_query(&mut self) {
        self.query_stack.push(self.query.clone());
    }

    fn pop_query(&mut self) {
        self.query = self.query_stack.pop().unwrap();
    }
}

#[cfg(test)]
mod test {
    use crate::compile;
    use crate::Options;

    #[test]
    fn test_end_with_new_line() {
        let sql = compile("from a", &Options::default().no_signature()).unwrap();
        assert_eq!(sql, "SELECT\n  *\nFROM\n  a\n")
    }
}
