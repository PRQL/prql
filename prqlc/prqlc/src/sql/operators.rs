use std::collections::HashMap;
use std::iter::zip;
use std::path::PathBuf;
use std::sync::OnceLock;

use itertools::Itertools;

use super::gen_expr::{translate_operand, ExprOrSource, SourceExpr};
use super::{Context, Dialect};
use crate::ir::{decl, pl, rq};
use crate::utils::Pluck;
use crate::Result;
use crate::{debug, semantic};
use crate::{Error, WithErrorInfo};

fn std() -> &'static decl::Module {
    static STD: OnceLock<decl::Module> = OnceLock::new();
    STD.get_or_init(|| {
        let _suppressed = debug::log_suppress();

        let std_lib = crate::SourceTree::new(
            [(
                PathBuf::from("std.sql.prql"),
                include_str!("./std.sql.prql").to_string(),
            )],
            None,
        );
        let ast = crate::parser::parse(&std_lib).unwrap();
        let context = semantic::resolve(ast).unwrap();

        context.module
    })
}

pub(super) fn translate_operator_expr(expr: rq::Expr, ctx: &mut Context) -> Result<ExprOrSource> {
    let (name, args) = expr.kind.into_operator().unwrap();

    let source = translate_operator(name, args, ctx).with_span(expr.span)?;

    Ok(ExprOrSource::Source(source))
}

pub(super) fn translate_operator(
    name: String,
    args: Vec<rq::Expr>,
    ctx: &mut Context,
) -> Result<SourceExpr> {
    let (operator_impl, binding_strength, window_frame, coalesce) =
        find_operator_impl(&name, ctx.dialect_enum).unwrap();
    let parent_binding_strength = binding_strength.unwrap_or(100);

    let args: HashMap<&str, _> = zip(operator_impl.params, args).collect();

    // body can only be an s-string
    let body = match &operator_impl.body.kind {
        pl::ExprKind::Literal(pl::Literal::Null) => {
            return Err(Error::new_simple(format!(
                "operator {} is not supported for dialect {}",
                name, ctx.dialect_enum
            )))
        }
        pl::ExprKind::SString(items) => items,
        _ => {
            return Err(Error::new_assert(
                "bad RQ operator implementation for SQL: expected function or a plain s-string",
            ))
        }
    };

    let mut text = String::new();

    for item in body {
        match item {
            pl::InterpolateItem::Expr { expr, format } => {
                // s-string exprs can only contain idents
                let ident = expr.kind.as_ident();
                let ident = ident.as_ref().unwrap();

                // lookup args
                let arg = args.get(ident.name.as_str()).unwrap().clone();

                // binding strength
                let required_strength = format
                    .as_ref()
                    .and_then(|f| f.parse::<i32>().ok())
                    .unwrap_or(parent_binding_strength);

                // translate args
                let arg = translate_operand(
                    arg,
                    false,
                    required_strength,
                    super::gen_expr::Associativity::Both,
                    ctx,
                )?;

                text += &arg.into_source();
            }
            pl::InterpolateItem::String(s) => {
                text += s;
            }
        }
    }

    let mut binding_strength = parent_binding_strength;

    if !ctx.query.window_function {
        if let Some(default) = coalesce {
            text = format!("COALESCE({text}, {default})");
            binding_strength = 100;
        }
    }

    Ok(SourceExpr {
        text,
        binding_strength,
        window_frame,
    })
}

struct OperatorImpl<'a> {
    body: &'a pl::Expr,
    params: Vec<&'a str>,
}

fn find_operator_impl(
    operator_name: &str,
    dialect: Dialect,
) -> Option<(OperatorImpl<'_>, Option<i32>, bool, Option<String>)> {
    let operator_name = operator_name.strip_prefix("std.").unwrap();
    let operator_ident = pl::Ident::from_path(
        operator_name
            .split('.')
            .map(String::from)
            .collect::<Vec<_>>(),
    );

    let dialect_module = std().get(&pl::Ident::from_name(dialect.to_string()));

    let mut impl_decl = None;

    if let Some(dialect_module) = dialect_module {
        let module = dialect_module.kind.as_module().unwrap();
        impl_decl = module.get(&operator_ident);
    }

    if impl_decl.is_none() {
        impl_decl = std().get(&operator_ident);
    }

    let impl_decl = impl_decl?;

    let impl_expr = impl_decl.kind.as_expr().unwrap();
    let operator_impl = match &impl_expr.kind {
        pl::ExprKind::Func(func) => {
            let params: Vec<_> = func
                .named_params
                .iter()
                .chain(func.params.iter())
                .map(|x| x.name.split('.').last().unwrap_or(x.name.as_str()))
                .collect();
            OperatorImpl {
                body: func.body.as_ref(),
                params,
            }
        }
        _ => OperatorImpl {
            body: impl_expr.as_ref(),
            params: Vec::new(),
        },
    };

    let annotation = impl_decl.annotations.iter().exactly_one().ok();
    let mut annotation = annotation
        .and_then(|x| into_tuple_items(*x.expr.clone()).ok())
        .unwrap_or_default();

    let binding_strength = pluck_annotation(&mut annotation, "binding_strength")
        .and_then(|literal| literal.into_integer().ok())
        .map(|int| int as i32);

    let window_frame = pluck_annotation(&mut annotation, "window_frame")
        .and_then(|literal| literal.into_boolean().ok())
        .unwrap_or_default();

    let coalesce =
        pluck_annotation(&mut annotation, "coalesce").and_then(|val| val.into_string().ok());

    Some((operator_impl, binding_strength, window_frame, coalesce))
}

fn pluck_annotation(
    annotation: &mut Vec<(String, pl::ExprKind)>,
    name: &str,
) -> Option<pl::Literal> {
    annotation
        .pluck(|(n, val)| if n == name { Ok(val) } else { Err((n, val)) })
        .into_iter()
        .next()
        .and_then(|val| val.into_literal().ok())
}

/// Find the items in a `@{a=b}`. We're only using annotations with tuples;
/// we can consider formalizing this constraint.
fn into_tuple_items(expr: pl::Expr) -> Result<Vec<(String, pl::ExprKind)>, pl::Expr> {
    match expr.kind {
        pl::ExprKind::Tuple(items) => items
            .into_iter()
            .map(|item| Ok((item.alias.clone().unwrap(), item.kind)))
            .collect(),
        _ => Err(expr),
    }
}
