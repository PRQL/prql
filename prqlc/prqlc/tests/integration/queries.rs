#![cfg(not(target_family = "wasm"))]
use std::path::Path;
use std::{env, fs};

use insta::assert_debug_snapshot;
use insta::{assert_snapshot, with_settings};

use prqlc::sql::Dialect;
use prqlc::{Options, Target};
use test_each_file::test_each_path;

mod lex {
    use super::*;

    test_each_path! { in "./prqlc/prqlc/tests/integration/queries" => run }

    fn run(prql_path: &Path) {
        let test_name = prql_path.file_stem().unwrap().to_str().unwrap();
        let prql = fs::read_to_string(prql_path).unwrap();

        let tokens = prqlc_parser::lex_source(&prql).unwrap();

        with_settings!({ input_file => prql_path }, {
            assert_debug_snapshot!(test_name, tokens)
        });
    }
}

mod compile {
    use super::*;

    // If this is giving compilation errors saying `expected identifier, found keyword`,
    // rename the filenames in queries to something that's not a keyword in Rust.
    test_each_path! { in "./prqlc/prqlc/tests/integration/queries" => run }

    fn run(prql_path: &Path) {
        let test_name = prql_path.file_stem().unwrap().to_str().unwrap();
        let prql = fs::read_to_string(prql_path).unwrap();
        if prql.contains("generic:skip") {
            return;
        }

        let target = Target::Sql(Some(Dialect::Generic));
        let options = Options::default().no_signature().with_target(target);

        let sql = prqlc::compile(&prql, &options).unwrap();

        with_settings!({ input_file => prql_path }, {
            assert_snapshot!(test_name, &sql, &prql)
        });
    }
}

mod fmt {
    use super::*;

    test_each_path! { in "./prqlc/prqlc/tests/integration/queries" => run }

    fn run(prql_path: &Path) {
        let test_name = prql_path.file_stem().unwrap().to_str().unwrap();
        let prql = fs::read_to_string(prql_path).unwrap();

        let pl = prqlc::prql_to_pl(&prql).unwrap();
        let formatted = prqlc::pl_to_prql(&pl).unwrap();

        with_settings!({ input_file => prql_path }, {
            assert_snapshot!(test_name, &formatted, &prql)
        });
    }
}

#[cfg(any(feature = "test-dbs", feature = "test-dbs-external"))]
mod results {
    use super::*;

    use std::{ops::DerefMut, sync::Mutex};

    use once_cell::sync::Lazy;
    use prqlc::sql::SupportLevel;

    use crate::dbs::{ConnectionCfg, DbConnection, DbProtocol};

    static CONNECTIONS: Lazy<Mutex<Vec<DbConnection>>> = Lazy::new(init_connections);

    fn init_connections() -> Mutex<Vec<DbConnection>> {
        let configs = [
            ConnectionCfg {
                dialect: Dialect::SQLite,
                data_file_root: "tests/integration/data/chinook".to_string(),

                protocol: DbProtocol::SQLite,
            },
            ConnectionCfg {
                dialect: Dialect::DuckDb,
                data_file_root: "tests/integration/data/chinook".to_string(),

                protocol: DbProtocol::DuckDb,
            },
            ConnectionCfg {
                dialect: Dialect::Postgres,
                data_file_root: "/tmp/chinook".to_string(),

                protocol: DbProtocol::Postgres {
                    url: "host=localhost user=root password=root dbname=dummy".to_string(),
                },
            },
            ConnectionCfg {
                dialect: Dialect::MySql,
                data_file_root: "/tmp/chinook".to_string(),

                protocol: DbProtocol::MySql {
                    url: "mysql://root:root@localhost:3306/dummy".to_string(),
                },
            },
            ConnectionCfg {
                dialect: Dialect::ClickHouse,
                data_file_root: "chinook".to_string(),

                protocol: DbProtocol::MySql {
                    url: "mysql://default:@localhost:9004/dummy".to_string(),
                },
            },
            ConnectionCfg {
                dialect: Dialect::GlareDb,
                data_file_root: "/tmp/chinook".to_string(),

                protocol: DbProtocol::Postgres {
                    url: "host=localhost user=glaredb dbname=glaredb port=6543".to_string(),
                },
            },
            ConnectionCfg {
                dialect: Dialect::MsSql,
                data_file_root: "/tmp/chinook".to_string(),

                protocol: DbProtocol::MsSql,
            },
        ];

        let mut connections = Vec::new();
        for cfg in configs {
            if !matches!(
                cfg.dialect.support_level(),
                SupportLevel::Supported | SupportLevel::Unsupported
            ) {
                continue;
            }

            // The filtering is not a great design, since it doesn't proactively
            // check that we can get connections; but it's a compromise given we
            // implement the external_dbs feature using this.
            let Some(mut connection) = DbConnection::new(cfg) else {
                continue;
            };

            connection.setup();

            connections.push(connection);
        }
        Mutex::new(connections)
    }

    test_each_path! { in "./prqlc/prqlc/tests/integration/queries" => run }

    fn run(prql_path: &Path) {
        let test_name = prql_path.file_stem().unwrap().to_str().unwrap();
        let prql = fs::read_to_string(prql_path).unwrap();

        let data_file_root_keyword = "data_file_root";
        let is_contains_data_root = prql.contains(data_file_root_keyword);

        // for each of the connections
        let mut results = Vec::new();
        for con in CONNECTIONS.lock().unwrap().deref_mut() {
            if !con.should_run_query(&prql) {
                continue;
            }

            let mut prql = prql.clone();
            if is_contains_data_root {
                prql = prql.replace(data_file_root_keyword, &con.cfg.data_file_root);
            }

            let dialect = con.cfg.dialect;

            println!("Executing {test_name} for {dialect}");
            let rows = con.run_query(&prql).unwrap();

            // convert into ad-hoc CSV
            let result = rows
                .iter()
                .map(|r| r.join(","))
                .collect::<Vec<_>>()
                .join("\n");

            results.push((dialect, result));
        }

        // insta::allow_duplicates!, but with reporting of which two cases are
        // not matching.
        let ((first_dialect, first_text), others) = results.split_first().unwrap();

        with_settings!({ input_file => prql_path }, {
            assert_snapshot!(test_name, first_text, &prql)
        });

        for (dialect, text) in others {
            similar_asserts::assert_eq!(
                first_text,
                text,
                "{} {} {}",
                test_name,
                first_dialect,
                dialect
            );
        }
    }
}
