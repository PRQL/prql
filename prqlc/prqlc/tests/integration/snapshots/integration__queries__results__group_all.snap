---
source: prqlc/prqlc/tests/integration/queries.rs
expression: "# mssql:test\nfrom a=albums\ntake 10\njoin tracks (==album_id)\ngroup {a.album_id, a.title} (aggregate price = (sum tracks.unit_price | math.round 2))\nsort album_id\n"
input_file: prqlc/prqlc/tests/integration/queries/group_all.prql
---
1,For Those About To Rock We Salute You,9.9
2,Balls to the Wall,0.99
3,Restless and Wild,2.97
4,Let There Be Rock,7.92
5,Big Ones,14.85
6,Jagged Little Pill,12.87
7,Facelift,11.88
8,Warner 25 Anos,13.86
9,Plays Metallica By Four Cellos,7.92
10,Audioslave,13.86
